ui/dashboards/src/utils/panelUtils.ts

```diff
@@ -26,6 +27,133 @@ export function getYForNewRow(group: PanelGroupDefinition) {
   return newRowY;
 }
 
+type PanelGroupItemBounds = {
+  /**
+   * Left horizontal position.
+   */
+  x1: number;
+  /**
+   * Right horizontal position.
+   */
+  x2: number;
+  /**
+   * Top vertical position.
+   */
+  y1: number;
+  /**
+   * Bottom vertical position
+   */
+  y2: number;
+};
+
+function getPanelBounds({ x, y, w, h }: PanelGroupItemLayout): PanelGroupItemBounds {
+  return {
+    x1: x,
+    x2: x + w,
+    y1: y,
+    y2: y + h,
+  };
+}
+
+export type PartialPanelGroupItemLayout = Omit<PanelGroupItemLayout, 'x' | 'y'>;
+
+/**
+ * Inserts a new panel into the layout with placement determined by a specified
+ * reference panel. The new panel is placed:
+ * - To the right of the reference panel if there is space available without
+ *   moving other panels.
+ * - Otherwise, directly below the reference panel. If other panels are below
+ *   this location, they will also shift downward because the grid uses
+ *   vertical-based compacting.
+ *
```

Comment:
I left this logic relatively generic around panel insertion (instead of very focused on duplication) because we may want to use it in the future for other functionality like copy-paste.

Reply:
Yeah I like this, another possible use case is if we add custom actions like 'Add to Native Dashboard' from other parts of our product internally.